examples/pytorch/seal/main.py

```diff
@@ -0,0 +1,175 @@
+import dgl
+from utils import parse_arguments
+from tqdm import tqdm
+from dgl import NID, EID
+from torch.nn import BCEWithLogitsLoss
+from utils import load_ogb_dataset, evaluate_hits
+import torch
+from sampler import SEALData, SEALDataLoader
+from model import GCN, DGCNN
+import numpy as np
+from logger import LightLogging
+import time
+
+
+def train(model, dataloader, loss_fn, optimizer, device, num_graphs=32, total_graphs=None):
+    model.train()
+
+    total_loss = 0
+    for g, labels in tqdm(dataloader, ncols=100):
+        g = g.to(device)
+        labels = labels.to(device)
+        optimizer.zero_grad()
+        logits = model(g, g.ndata['z'], g.ndata[NID], g.edata[EID])
+        loss = loss_fn(logits, labels)
+        loss.backward()
+        optimizer.step()
+        total_loss += loss.item() * num_graphs
+
+    return total_loss / total_graphs
+
+
+@torch.no_grad()
+def evaluate(model, dataloader, device):
+    model.eval()
+
+    y_pred, y_true = [], []
+    for g, labels in tqdm(dataloader, ncols=100):
+        g = g.to(device)
+        logits = model(g, g.ndata['z'], g.ndata[NID], g.edata[EID])
+        y_pred.append(logits.view(-1).cpu())
+        y_true.append(labels.view(-1).cpu().to(torch.float))
+
+    y_pred, y_true = torch.cat(y_pred), torch.cat(y_true)
+    pos_pred = y_pred[y_true == 1]
+    neg_pred = y_pred[y_true == 0]
+
+    return pos_pred, neg_pred
+
+
+def main(args, print_fn=print):
+    print_fn("Experiment arguments: {}".format(args))
+
+    if args.random_seed:
+        torch.manual_seed(args.random_seed)
+    else:
+        torch.manual_seed(123)
+    # Load dataset
+    if args.dataset.startswith('ogbl'):
+        graph, split_edge = load_ogb_dataset(args.dataset)
+    else:
+        raise NotImplementedError
+
+    num_nodes = graph.num_nodes()
+
+    # set gpu
+    if args.gpu_id >= 0 and torch.cuda.is_available():
+        device = 'cuda:{}'.format(args.gpu_id)
+    else:
+        device = 'cpu'
+
+    if args.dataset == 'ogbl-collab':
+        # ogbl-collab dataset is multi-edge graph
+        use_coalesce = True
+    else:
+        use_coalesce = False
+
+    # Generate positive and negative edges and corresponding labels
+    # Sampling subgraphs and generate node labeling features
+    seal_data = SEALData(g=graph, split_edge=split_edge, hop=args.hop, neg_samples=args.neg_samples,
+                         subsample_ratio=args.subsample_ratio, use_coalesce=use_coalesce, prefix=args.dataset,
+                         save_dir=args.save_dir, num_workers=args.num_workers, print_fn=print_fn)
+    node_attribute = seal_data.ndata['feat']
+    edge_weight = seal_data.edata['edge_weight'].float()
+
+    train_data = seal_data('train')
+    val_data = seal_data('valid')
+    test_data = seal_data('test')
+
+    train_graphs = len(train_data.graph_list)
+
+    # Set data loader
```

Comment:
The model will be out with 0.6 release. In that case we may only need the GraphDataLoader solution.